ddtrace/internal/processor/trace.py

```diff
@@ -0,0 +1,127 @@
+import abc
+from collections import defaultdict
+import threading
+from typing import DefaultDict
+from typing import Iterable
+from typing import List
+from typing import Optional
+
+import attr
+import six
+
+from ddtrace.internal.logger import get_logger
+from ddtrace.internal.processor import SpanProcessor
+from ddtrace.internal.writer import TraceWriter
+from ddtrace.span import Span
+
+
+log = get_logger(__name__)
+
+
+@attr.s
+class TraceProcessor(six.with_metaclass(abc.ABCMeta)):
+    def __attrs_post_init__(self):
+        # type () -> None
+        """Default post initializer which logs the representation of the
+        TraceProcessor at the ``logging.DEBUG`` level.
+
+        The representation can be modified with the ``repr`` argument to the
+        attrs attribute::
+
+            @attr.s
+            class MyTraceProcessor(TraceProcessor):
+                field_to_include = attr.ib(repr=True)
+                field_to_exclude = attr.ib(repr=False)
+        """
+        log.debug("initialized trace processor %r", self)
+
+    @abc.abstractmethod
+    def process_trace(self, trace):
+        # type: (List[Span]) -> Optional[List[Span]]
+        """Processes a trace.
+
+        None can be returned to prevent the trace from being exported.
+        """
+        pass
+
+
+@attr.s
+class TraceSamplingProcessor(TraceProcessor):
+    def process_trace(self, trace):
+        # type: (List[Span]) -> Optional[List[Span]]
+        sampled_spans = [s for s in trace if s.sampled]
+        if len(sampled_spans) == 0:
+            log.info("dropping trace, %d spans unsampled", len(trace))
+            return None
+
+        log.info("trace %d sampled (%d/%d spans sampled)", trace[0].trace_id, len(sampled_spans), len(trace))
+        return trace
+
+
+@attr.s
+class SpansToTraceProcessor(SpanProcessor):
+    """Processor that aggregates spans together by trace_id and submits the
+    finalized trace if the trace is assumed to be complete[0] or if a partial
+    flushing threshold has been met[1].
+
+    [0] A trace is assumed to be complete if all the spans that have been
+    created with the trace_id have finished. When this condition is met the
+    completed spans are returned.
+
+    [1] A minimum threshold of spans can be specified with the
+    ``partial_flush_min_spans`` argument. When this threshold of spans has been
+    finished the finished spans of the trace are returned.
+    """
+
+    @attr.s
+    class _Trace(object):
```

Comment:
Wondering whether we could actually do without this intermediate object and just add the fields to the parent class

Reply:
🤔 like keep 2 separate dictionaries of traceid->spans and traceid->num_finished? I think this was the case originally but we opted to group the data in a class